sql: TestSchemaUniqueColumnDropFailure failed

sql.TestSchemaUniqueColumnDropFailure [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4743093&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4743093&tab=artifacts#/) on release-22.1 @ [a88367d7bd168a3614b7946cd7b77bb9c6374a30](https://github.com/cockroachdb/cockroach/commits/a88367d7bd168a3614b7946cd7b77bb9c6374a30):


```
=== RUN   TestSchemaUniqueColumnDropFailure
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestSchemaUniqueColumnDropFailure2627273156
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:2309: -- test log scope end --
    schema_changer_test.go:2309: Leaked goroutine: goroutine 47886225 [IO wait]:
        internal/poll.runtime_pollWait(0x7ffac2cd6388, 0x72)
        	GOROOT/src/runtime/netpoll.go:234 +0x89
        internal/poll.(*pollDesc).wait(0xc018673d80, 0xc018807360, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc018673d80, {0xc018807360, 0x1, 0x200})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc018673d80, {0xc018807360, 0x7af765, 0xc018673d80})
        	GOROOT/src/net/fd_posix.go:56 +0x29
        net.(*conn).Read(0xc00eadadc0, {0xc018807360, 0x7c400e, 0xc00eadadc0})
        	GOROOT/src/net/net.go:183 +0x45
        io.ReadAtLeast({0x5cb7080, 0xc00eadadc0}, {0xc018807360, 0x1, 0x200}, 0x1)
        	GOROOT/src/io/io.go:328 +0x9a
        io.ReadFull(...)
        	GOROOT/src/io/io.go:347
        github.com/lib/pq.(*conn).ssl(0xc018807340, 0xc00007e038)
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:1087 +0x145
        github.com/lib/pq.(*Connector).open(0xc015090288, {0x5d52fb8, 0xc00007e038})
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:327 +0x2b8
        github.com/lib/pq.DialOpen({0x5cd6638, 0xc0188793e0}, {0xc000a3d340, 0xc0000bffc0})
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:291 +0x7a
        github.com/lib/pq.Open(...)
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:281
        github.com/lib/pq.Driver.Open({}, {0xc000a3d340, 0x15c})
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:54 +0x7a
        database/sql.dsnConnector.Connect(...)
        	GOROOT/src/database/sql/sql.go:761
        database/sql.(*DB).conn(0xc012165c70, {0x5d52fb8, 0xc00007e038}, 0x1)
        	GOROOT/src/database/sql/sql.go:1364 +0x7ac
        database/sql.(*DB).exec(0xc002654718, {0x5d52fb8, 0xc00007e038}, {0x49aca73, 0x20}, {0x0, 0x0, 0x0}, 0x0)
        	GOROOT/src/database/sql/sql.go:1622 +0x5d
        database/sql.(*DB).ExecContext(0x10585e0, {0x5d52fb8, 0xc00007e038}, {0x49aca73, 0x20}, {0x0, 0x0, 0x0})
        	GOROOT/src/database/sql/sql.go:1601 +0xed
        database/sql.(*DB).Exec(...)
        	GOROOT/src/database/sql/sql.go:1618
        github.com/cockroachdb/cockroach/pkg/sql_test.TestSchemaUniqueColumnDropFailure.func4()
        	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/schema_changer_test.go:2283 +0x45
        created by github.com/cockroachdb/cockroach/pkg/sql_test.TestSchemaUniqueColumnDropFailure
        	github.com/cockroachdb/cockroach/pkg/sql_test/pkg/sql/schema_changer_test.go:2281 +0x685
--- FAIL: TestSchemaUniqueColumnDropFailure (6.59s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemaUniqueColumnDropFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-14577

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -2262,6 +2262,8 @@ func TestSchemaUniqueColumnDropFailure(t *testing.T) {
 		},
 		JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
 	}
+	var wg sync.WaitGroup
+	defer wg.Wait()
 	server, sqlDB, kvDB := serverutils.StartServer(t, params)
 	defer server.Stopper().Stop(context.Background())
 
@@ -2282,7 +2284,9 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT UNIQUE DEFAULT 23 CREATE FAMILY F3
 	}
 
 	// A schema change that fails.
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		// This query stays blocked until the end of the test.
 		_, _ = sqlDB.Exec(`ALTER TABLE t.test DROP column v`)
 	}()

